README.rst
django-initial-avatars

======================

.. image:: https://badge.fury.io/py/django-initial-avatars.svg  
    :target: https://badge.fury.io/py/django-initial-avatars
.. image:: https://travis-ci.org/axiome-oss/django-initial-avatars.svg?branch=master
    :target: https://travis-ci.org/axiome-oss/django-initial-avatars

django-initial-avatars is a simple Django app which generates avatars based on username and initials. If django_gravatar is installed, user's gravatar is preferred.

Examples
-----------

* API endpoint

.. image:: https://metod-site.s3.amazonaws.com/media/25/initial_avatars.png
    :target: http://www.metod.io/fr/blog/2015/12/02/release-django-initial-avatars/
    :alt: example of django-initial-avatars on Metod
    
* Template tag

.. image:: https://metod-site.s3.amazonaws.com/media/25/initial_avatars_email.png
    :target: http://www.metod.io/fr/blog/2015/12/02/release-django-initial-avatars/
    :alt: example of django-initial-avatars in Metod emails

Dependencies
------------

Generating avatars requires a `Pillow` installation with `freetype` support.

``freetype`` can easily be installed on ubuntu with::
	
	$ sudo aptitude install libfreetype6-dev

or on OS X with `homebrew`::

    $ brew install freetype

Make sure the following packet are installed on your system to enable PNG and JPG support on Pillow::

    $ sudo aptitude install libjpeg-dev zlib1g-dev libpng12-dev

Pillow may need to be rebuilt after installing the libraries.

Font licensing
--------------

The font ``Ubuntu Monospace`` is used to generate the avatar.
The font is licensed under the Ubuntu Font Licence, see the
`License <http://font.ubuntu.com/licence/>`_

Quick start
-----------
1. install app requirements

2. install django-initial-avatars through pip::

    pip install django-initial-avatars

3. If you want to use gravatar for users who have one, install django-gravatar2::

    pip install django-gravatar2

4. Add "django-initial-avatars" to your INSTALLED_APPS setting like this::

    INSTALLED_APPS = (
        ...
        'initial_avatars',
        ['django_gravatar',]
    )

5. Include the django-initial-avatar URLconf in your project urls.py like this::

    url(r'^avatar/', include('initial_avatars.urls')),

6. Launch development server::

	python manage.py runserver

7. Each user has now an endpoint for his avatar::

	localhost:8000/avatar/user_id/[size/]

8. In your templates, use::

    {% load initialavatar %}

    {% get_initial_avatar user [size] ['shape'] %}

    {% get_initial_avatar user.email [size] ['shape'] %}


Settings
-----------

A few settings are at your disposal

* AVATAR_STORAGE_BACKEND allows you to use a custom storage backend instead of the default one::

    AVATAR_STORAGE_BACKEND = 'myproject.custom_storages.AvatarStorage'

* AVATAR_STORAGE_FOLDER allows you to customize the root folder on the storage backend, default to 'avatars'::

    AVATAR_STORAGE_FOLDER = 'myfolder'

* AVATAR_DEFAULT_SHAPE allows you to choose the default shape of the image, possible options are 'circle' or 'square', default to 'square'

  More shapes can be easily addded, just open an issue on github::

    AVATAR_DEFAULT_SHAPE = 'circle'

* GRAVATAR_DEFAULT_SIZE allows you to choose the default size of the image, setting name used for compatibility with django_gravatar, default to '80'::

    GRAVATAR_DEFAULT_SIZE = 100


Experimental settings
------------
Those feature are available but miss tests, contributions would be appreciated :)

* AVATAR_HIGH_RESOLUTION can be used to display retina ready avatas, default to False::

    AVATAR_HIGH_RESOLUTION = True

* AVATAR_COLORS can be used to randomly choose a color from a tuple of pre-defined colors at first avatar generation, no default::

    AVATAR_COLORS = ((37, 114, 221), (26, 193, 255),)

* AVATAR_DEFAULT_FOREGROUND can be used to define a default color to the foreground, 'black' and 'white' available, no default. I advise not to use it without AVATAR_COLORS::

    AVATAR_DEFAULT_FOREGROUND = 'white'

* AVATAR_GENERATOR_BACKEND can be used to extend the avatar generator and adjust it to your needs, especially for font customization. Refer to initial_avatars/generator.py for more information::

    AVATAR_GENERATOR_BACKEND = 'my_project.avatar_backend.MyAvatarBackend'


.. code-block:: python
    from initial_avatars.generator import AvatarGenerator
    from PIL import ImageFont
    import os
    class MyAvatarBackend(AvatarGenerator):
         def font(self):
            font_path = '/path/to/your/font'
            font_size = self.font_size()
            return ImageFont.truetype(font_path, size=font_size)



Tests
--------------

Django-initial-avatars is provided with tests, they require django-gravatar2 and tox

You can launch them in the virtualenv like this::

        tox

It might happen that a calculated position fails because of a minor difference in the result, don't care about it.

Contributions
--------------

Contributions are welcome ! Feel free to write an issue for any feedback you have or send a pull request on `Github <https://github.com/axiome-oss/django-initial-avatars>`_

Used on
--------------

* `Metod <http://www.metod.io/>`_
* Add your website here !


changelog
v0.9:
	- add compatibility with retina screens 'AVATAR_HIGH_RESOLUTION' settings
	- add 'AVATAR_COLORS' settings to choose avatar colors from defined list
	- add 'AVATAR_DEFAULT_FOREGROUND' settings to use black or white as default letter color
	- add possibility to custom font with 'AVATAR_GENERATOR_BACKEND' settings
	- django 1.10 compatible
	- thanks to thomashamlin :
		- improve performance in templates
		- add support for last names that start with lowercase characters like "da Rocha"

v0.8.0:
    - add possibility to use cirle avatars with 'AVATAR_DEFAULT_SHAPE' settings (default to 'square')
    - add possibility to customize root folder on with 'AVATAR_STORAGE_FOLDER' settings (default to 'avatars')
    - changed file naming
    - improved image quality
    - update documentation
    - python 3.4 compatible
    - django 1.9 compatible
    - use tox for unittesting

v0.7.5:
	- really fix bug with accent in username, add tests for this case

v0.7.4:
	- add # -*- coding: utf-8 -*- at beginning of files to prevent bug 
	when username has accent

v0.7.3: 
	- removes the last_modified_func decorator because of poor
	performances

v0.7.2:
	- removes tenant_schema specific code

v0.7.1:
	- fix: last_modification on view did not change when avatar changed on
	storage backend

v0.7:
	- add a setting to let developers choose a specific backend for avatar storage

v0.6:
	- use StringIO and django InMemoryUploadedFile instead of tmp file
	- update README

v0.5.2:
	- update README
	- add doc string to generator.py
	- add changelog

v0.5.1:
	- fix: missing import when using django-tenant-schemas

v0.5.0:
	- security: use user_id instead of user_username for the url

v0.4:
	- feat: avatar path now use md5 of username + first_name + last_name insted of simple username:
		permit to updates the avatar when user changes its settings
		+ better for security


initial_avatars/generator.py
# -*- coding: utf-8 -*-
from __future__ import division
from __future__ import unicode_literals
import os
import re
try:
    from django_gravatar.helpers import get_gravatar_url, has_gravatar
except ImportError:
    pass
from django.utils.html import escape
from django.utils import timezone
from django.conf import settings
from django.core.files.storage import default_storage, get_storage_class
from django.core.files.uploadedfile import InMemoryUploadedFile
from PIL import Image, ImageDraw, ImageFont
from io import BytesIO
from math import sqrt
from hashlib import md5
from datetime import datetime
from random import choice
from .utils import AVATAR_SHAPE_SETTINGS, AVATAR_FOREGROUND_COLORS, AvatarShapeException, AvatarForegroundColorException
from .models import Background
from .compat import urlopen

GRAVATAR_DEFAULT_SIZE = getattr(settings, 'GRAVATAR_DEFAULT_SIZE', 80)
AVATAR_SHAPE = getattr(settings, 'AVATAR_DEFAULT_SHAPE', 'square')
AVATAR_STORAGE_FOLDER = getattr(settings, 'AVATAR_STORAGE_FOLDER', 'avatars')
AVATAR_HIGH_RESOLUTION = getattr(settings, 'AVATAR_HIGH_RESOLUTION', False)
AVATAR_COLORS = getattr(settings, 'AVATAR_COLORS', False)

try:
    AVATAR_DEFAULT_FOREGROUND = AVATAR_FOREGROUND_COLORS[settings.AVATAR_DEFAUL_TEXT_COLOR]
except KeyError:
    raise AvatarForegroundColorException
except AttributeError:
    pass

try:
    AVATAR_STORAGE_BACKEND = get_storage_class(settings.AVATAR_STORAGE_BACKEND)()
except AttributeError:
    AVATAR_STORAGE_BACKEND = default_storage


class AvatarGenerator(object):
    """
        inspired by https://github.com/4teamwork/ftw.avatar
    """

    def __init__(self, user, size=GRAVATAR_DEFAULT_SIZE, shape=AVATAR_SHAPE):
        self.user = user
        self.work_size = size * 10
        self.size = size
        self.shape = shape
        try:
            self.image_format = AVATAR_SHAPE_SETTINGS[shape]['image_format']
            self.content_type = AVATAR_SHAPE_SETTINGS[shape]['content_type']
        except KeyError:
            raise AvatarShapeException
        self.css_class = None
        self.url = None
        if AVATAR_HIGH_RESOLUTION:
            self.high_res_url = None

    def name(self, high_res=False):
        """
            returns the name of the img file
        """
        if high_res:
            name = "{0}x{0}_{1}@2x.{2}".format(self.size, self.shape, self.image_format)
        else:
            name = '{0}x{0}_{1}.{2}'.format(self.size, self.shape, self.image_format)
        return name

    def path(self, high_res=False):
        """
            returns the path of the img file
        """
        user_hash = md5(os.path.join(self.user.username, self.user.first_name, self.user.last_name).encode('utf-8')).hexdigest()
        user_path = os.path.join(AVATAR_STORAGE_FOLDER, user_hash, self.name(high_res))
        return user_path

    def font_size(self):
        """
            returns the size of the font calculated according to the size of requested avatar
        """
        font_size = int(self.work_size * (1 - 0.1 * len(self.text())))
        return font_size

    def font(self):
        """
            returns an ImageFont object with the font used to generate the avatar
        """
        font_path = os.path.join(os.path.dirname(__file__), 'font', 'UbuntuMono-B.ttf')
        font_size = self.font_size()
        return ImageFont.truetype(font_path, size=font_size)

    def choose_random_background(self):
        try:
            backgroundObject = Background.objects.get(user=self.user)
            background = (backgroundObject.R, backgroundObject.G, backgroundObject.B)
        except Background.DoesNotExist:
            background = choice(AVATAR_COLORS)
            Background.objects.create(user=self.user, R=background[0], G=background[1], B=background[2])
        return background

    def background(self):
        """
            returns the background color based on the username md5
        """
        if AVATAR_COLORS:
            background = self.choose_random_background()
        else:
            hash = md5(self.user.username.encode('utf-8')).hexdigest()
            hash_values = (hash[:8], hash[8:16], hash[16:24])
            background = tuple(int(value, 16) % 256 for value in hash_values)
        return background

    def brightness(self):
        """
            returns the brightness of the background color
            explanation of the formula on http://www.nbdtech.com/Blog/archive/2008/04/27/Calculating-the-Perceived-Brightness-of-a-Color.aspx
        """
        rCoef = 0.241
        gCoef = 0.691
        bCoef = 0.068
        background = self.background()
        brightness = sqrt(rCoef * background[0]**2 + gCoef * background[1]**2 + bCoef * background[2]**2)
        return brightness

    def foreground(self):
        """
            returns black or white according to the brightness
        """
        try:
            return AVATAR_DEFAULT_FOREGROUND
        except NameError:
            brightness = self.brightness()
            if brightness > 130:
                return (0, 0, 0)
            else:
                return (255, 255, 255)

    def position(self, draw):
        """
            returns the position where the initials must be printed
        """
        text_width, text_height = draw.textsize(self.text(), font=self.font())
        left = ((self.work_size - text_width) / 2)
        top = ((self.work_size - text_height) / 4)
        return left, top

    def text(self):
        """
            returns the text to be printed on the avatar
            first letter of first_name and last_name if they exists
            first letter of the username if not

            Attempts to find first capital letter of last name. This supports
            names such as "Alexandre da Rocha" by using the correct form "AR".
        """
        if self.user.first_name and self.user.last_name:
            initial = self.user.first_name[:1].upper()

            match = re.search(r'[A-Z]', self.user.last_name)
            if match:
                # Use first capital letter
                initial += match.group()
            else:
                # No capitals found; just use first letter
                initial += self.user.last_name[:1].upper()
        else:
            initial = self.user.username[:1].upper()
        return initial

    def last_modification(self):
        """
            returns the avatar last_modification
        """
        try:
            if has_gravatar(self.user.email):
                info = urlopen(get_gravatar_url(email=self.user.email, size=self.size)).info()
                return datetime.strptime(info['Last-Modified'], "%a, %d %b %Y %H:%M:%S GMT")
        except NameError:
            pass
        if AVATAR_STORAGE_BACKEND.exists(self.path()):
            try:
                return AVATAR_STORAGE_BACKEND.modified_time(self.path())
            except AttributeError:
                return timezone.now()
        else:
            return None

    def genavatar(self):
        if self.shape == 'square':
            return self.gen_image_avatar(self.background())
        elif self.shape == 'circle':
            return self.gen_image_avatar((255, 0, 0, 0))
        else:
            raise AvatarShapeException

    def gen_image_avatar(self, background):
        work_image = Image.new('RGBA', (self.work_size, self.work_size), background)
        draw = ImageDraw.Draw(work_image)
        if self.shape == 'circle':
            draw.ellipse((0, 0, self.work_size, self.work_size), fill=self.background())
        w, h = self.position(draw)
        draw.text((w, h), self.text(), fill=self.foreground(), font=self.font())
        image = work_image.resize((self.size, self.size), resample=Image.BILINEAR)
        self.url = self.save_avatar(image)
        if AVATAR_HIGH_RESOLUTION:
            high_res_image = work_image.resize((self.size * 2, self.size * 2), resample=Image.BILINEAR)
            self.high_res_url = self.save_avatar(high_res_image, high_res=True).decode('utf-8')
        return self.url

    def save_avatar(self, image, high_res=False):
        image_io = BytesIO()
        image.save(image_io, format=self.content_type)
        image_io.seek(0, os.SEEK_END)
        image_io_length = image_io.tell()
        image_name = self.name(high_res)
        try:
            django_file = InMemoryUploadedFile(image_io, None, image_name, 'image/{0}'.format(self.content_type.lower()), image_io_length, None)
            AVATAR_STORAGE_BACKEND.save(self.path(high_res), django_file)
            return AVATAR_STORAGE_BACKEND.url(self.path(high_res))
        except Exception as e:
            raise e

    def get_avatar_url(self):
        """
            returns the url of the avatar on the storage backed
        """
        try:
            if has_gravatar(self.user.email):
                self.css_class = "gravatar"
                self.url = escape(get_gravatar_url(email=self.user.email, size=self.size))
                return self.url
        except NameError:
            pass

        self.css_class = "initial-avatar"
        if AVATAR_STORAGE_BACKEND.exists(self.path()):
            self.url = AVATAR_STORAGE_BACKEND.url(self.path())
        else:
            self.url = self.genavatar()
        if AVATAR_HIGH_RESOLUTION:
            self.high_res_url = AVATAR_STORAGE_BACKEND.url(self.path(high_res=True))
            return self.url, self.high_res_url
        return self.url

    def get_avatar(self):
        """
            returns an html img tag with the avatar
        """
        self.get_avatar_url()
        if AVATAR_HIGH_RESOLUTION:
            return '<img class="{css_class}" src="{src}" srcset="{src} 1x, {high_res_src} 2x" width="{width}" height="{height}"/>'.format(
                css_class=self.css_class,
                src=self.url,
                high_res_src=escape(self.high_res_url),
                width=self.size,
                height=self.size
            )
        return '<img class="{css_class}" src="{src}" width="{width}" height="{height}"/>'.format(
            css_class=self.css_class,
            src=self.url,
            width=self.size,
            height=self.size
        )
